lib/at_cmd_parser/at_params.c

```diff
@@ -29,6 +29,9 @@ static void at_param_clear(struct at_param *param)
 	__ASSERT(param != NULL, "Parameter cannot be NULL.");
 
 	if ((param->type == AT_PARAM_TYPE_STRING) ||
+	    (param->type == AT_PARAM_TYPE_CMD_SET) ||
```

Comment:
For set, read and test, the value is a normal String value. We already offer this type of value. Why adding a new type?
Instead, could you use another field/enum to indicates if it is a set, read or test string?

Reply:
I need to carry the information about the type of command being processed - in this approach I confined this information in the param type of the first parameter.

But indeed, giving it a second thought, it might actually be an overuse to include this information in the parameter type, and indeed a better option would be to introduce a separate parameter for this case.

So what now I am thinking of is to provide an enumeration type:
```
enum at_cmd_type {
    AT_CMD_TYPE_SET_COMMAND,
    AT_CMD_TYPE_READ_COMMAND,
    AT_CMD_TYPE_TEST_COMMAND,
    AT_CMD_TYPE_NOTIFICATION,
}
```
and then, within the `struct at_param_list` add a new field: `enum at_cmd_type cmd_type;`.

This approach should be way less intrusive and actually neater IMO. What do you think?

Reply:
I like this approach better. I do not think we need a new type.
Do we really need to store the command type in the parameter? All we need to know is that it is a String.
Can we have set and read parameters in the same list for instance? I think this information should only be available for the parser and is not related to parameters.
I looked at your updated test case, but I am not sure what the requirements really are.

Reply:
Ok, let me explain.

We have a project called `serial_lte_modem` which extends the list of available AT commands with a set of additional commands (let's call them `custom` as `proprietary` is already used by Nordic AT commands specification). This project uses `at_cmd_parser` module to extract the actual command parameters.

So far, as we have no support for READ and TEST commands, the sample analyzed the command string manually to determine the type of command it deals with and then used the `at_cmd_parser` to extract the command parameters only for the SET command (READ/TEST carry no parameters other than the command itself). This approach is cumbersome though, as it expects the `at_cmd_parser` user to understand the AT command syntax and actually requires to parse part of the command twice.

This PR intends to facilitate the process so that users of `at_cmd_parser` module can simply parse the command string and get all the information needed. Let's consider the custom `at#xsocket` command, it can have one of three formats (which complies to the AT command specification):
1) SET: `at#xsocket=1,1`
2) READ: `at#xsocket?`
3) TEST: `at#xsocket=?`

Running the `at_cmd_parser` over these command strings will give the following results if we ignore the command type (I've specified the param type in the parentheses):
1) `"at#xsocket"(string) 1(short) 1(short)`
2) `"at#xsocket"(string)`
3) `"at#xsocket"(string)`

As you can see, based only on the existing parameter types I can tell no difference between 2) and 3). If the actual command accepted SET command w/o parameters, 1) would be problematic as well. So we need extra information, informing the user what type of command was just parsed.

The original idea was to store the command type within the first parameter type, holding the actual command string, but we've given up on this. The current idea is to store the command type within the `struct at_param_list` as a separate field. There's no use providing this information at per-parameter basis, as it will not differ across the parameters (it's the command feature, not a paramter feature). The user could then easily access this information with a getter function, after the command was parsed, w/o diving into command internals.

Reply:
Thank you for the information.
We agree that adding a type for each parameter is not what we need.

I do not think the AT params list needs to know if the AT string is a set, test or read command. This information is known by the parser, so it should be stored in the parser, not in the parameter list.
What about adding an output parameter in the parser API?

```
int at_parser_params_from_str(const char *at_params_str, char **next_param_str,
    struct at_param_list *const list,
    const enum *const at_cmd_type);
```

Reply:
If you prefer to modify the API and provide command type as a separate output parameter, sure I can do it. Just note this will make the PR much larger as I will need to align all of the existing API calls.

I'm still not convinced though it's the right way to go. `struct at_param_list` is nothing more than a dissected AT command/notification and I see no harm in extending it with additional parser products. But I will not push here, it's your (as IoT team) module after all.